File: apps/web/lib/blueprint.ts
import type { TileSize } from "@/lib/types";

/**
 * TypeScript mirror of the widget view-blueprint contract.
 *
 * Authoritative source: `schemas/widget-blueprint.schema.json` and the Pydantic
 * models in `apps/api/gremlinboard_api/schemas/blueprint.py`. Keep this in sync
 * with that contract; everything here is intentionally permissive on read so the
 * universal renderer degrades gracefully instead of crashing on malformed data.
 */

export type BlueprintGap = "none" | "sm" | "md";
export type BlueprintStatusColor = "critical" | "warn" | "ok" | "neutral";
export type BlueprintStatusMap = Record<string, BlueprintStatusColor>;
export type BlueprintTier = "compact" | "medium" | "wide" | "tall" | "large";

export interface BlueprintShowIf {
  path: string;
  op: "exists" | "eq" | "gt" | "lt";
  value?: unknown;
}

interface BlueprintNodeBase {
  show_if?: BlueprintShowIf;
}

export interface StackNode extends BlueprintNodeBase {
  type: "stack";
  gap?: BlueprintGap;
  children: BlueprintNode[];
}

export interface RowNode extends BlueprintNodeBase {
  type: "row";
  gap?: BlueprintGap;
  children: BlueprintNode[];
}

export interface GridNode extends BlueprintNodeBase {
  type: "grid";
  gap?: BlueprintGap;
  columns: number;
  children: BlueprintNode[];
}

export interface ScrollNode extends BlueprintNodeBase {
  type: "scroll";
  gap?: BlueprintGap;
  children: BlueprintNode[];
}

export interface StatNode extends BlueprintNodeBase {
  type: "stat";
  label: string;
  value_path: string;
  unit?: string;
  emphasis?: "primary" | "secondary";
  trend_path?: string;
  status_path?: string;
  status_map?: BlueprintStatusMap;
}

export interface TextNode extends BlueprintNodeBase {
  type: "text";
  value_path?: string;
  literal?: string;
  variant: "title" | "body" | "caption" | "mono";
}

export interface BadgeItem {
  label_path?: string;
  literal?: string;
  status_path?: string;
  status_map?: BlueprintStatusMap;
}

export interface BadgeRowNode extends BlueprintNodeBase {
  type: "badge_row";
  items: BadgeItem[];
}

export interface ListItem {
  primary_path: string;
  secondary_path?: string;
  meta_path?: string;
  status_path?: string;
  status_map?: BlueprintStatusMap;
}

export interface ListNode extends BlueprintNodeBase {
  type: "list";
  items_path: string;
  limit?: number;
  item: ListItem;
}

export interface TableColumn {
  header: string;
  value_path: string;
  align?: "left" | "center" | "right";
}

export interface TableNode extends BlueprintNodeBase {
  type: "table";
  items_path: string;
  limit?: number;
  columns: TableColumn[];
}

export interface KeyValueEntry {
  label: string;
  value_path: string;
}

export interface KeyValueNode extends BlueprintNodeBase {
  type: "key_value";
  entries?: KeyValueEntry[];
  entries_path?: string;
}

export interface ProgressNode extends BlueprintNodeBase {
  type: "progress";
  value_path: string;
  max_path?: string;
  max_literal?: number;
  label?: string;
}

export interface SparklineNode extends BlueprintNodeBase {
  type: "sparkline";
  values_path: string;
  label?: string;
}

export interface TimerNode extends BlueprintNodeBase {
  type: "timer";
  target_path: string;
  direction: "down" | "up";
  label_path?: string;
}

export interface EmptyStateNode extends BlueprintNodeBase {
  type: "empty_state";
  message: string;
  show_if_empty_path: string;
}

export type LayoutNode = StackNode | RowNode | GridNode | ScrollNode;
export type PrimitiveNode =
  | StatNode
  | TextNode
  | BadgeRowNode
  | ListNode
  | TableNode
  | KeyValueNode
  | ProgressNode
  | SparklineNode
  | TimerNode
  | EmptyStateNode;
export type BlueprintNode = LayoutNode | PrimitiveNode;

export interface BlueprintLayouts {
  compact?: BlueprintNode;
  medium: BlueprintNode;
  wide?: BlueprintNode;
  tall?: BlueprintNode;
  large?: BlueprintNode;
}

export interface Blueprint {
  blueprint_version: "1";
  widget_id: string;
  layouts: BlueprintLayouts;
  defaults?: Record<string, BlueprintNode>;
}

export const EM_DASH = "—";

const LAYOUT_TYPES = new Set(["stack", "row", "grid", "scroll"]);

export function isLayoutNode(node: BlueprintNode): node is LayoutNode {
  return LAYOUT_TYPES.has(node.type);
}

/**
 * Safe dot-path resolver. Accepts paths of the form `a.b[0].c` and returns the
 * resolved value or `undefined` — it never throws, whatever the shape of state.
 */
export function resolvePath(state: unknown, path: string): unknown {
  if (typeof path !== "string" || path.length === 0) {
    return undefined;
  }

  let current: unknown = state;
  for (const segment of path.split(".")) {
    const match = /^([A-Za-z_][A-Za-z0-9_]*)((?:\[[0-9]+\])*)$/.exec(segment);
    if (!match) {
      return undefined;
    }

    const key = match[1];
    if (current === null || typeof current !== "object") {
      return undefined;
    }
    current = (current as Record<string, unknown>)[key];

    const indexTokens = match[2].match(/\[[0-9]+\]/g);
    if (indexTokens) {
      for (const token of indexTokens) {
        const index = Number(token.slice(1, -1));
        if (!Array.isArray(current) || index < 0 || index >= current.length) {
          return undefined;
        }
        current = current[index];
      }
    }
  }

  return current;
}

function toComparableNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

/** Evaluate a `show_if` guard against the widget state. Missing guard => visible. */
export function evaluateShowIf(state: unknown, showIf: BlueprintShowIf | undefined): boolean {
  if (!showIf) {
    return true;
  }

  const actual = resolvePath(state, showIf.path);
  switch (showIf.op) {
    case "exists":
      return actual !== undefined && actual !== null;
    case "eq":
      if (actual === showIf.value) {
        return true;
      }
      // Fall back to loose numeric comparison so 1 == "1" behaves intuitively.
      {
        const a = toComparableNumber(actual);
        const b = toComparableNumber(showIf.value);
        return a !== null && b !== null && a === b;
      }
    case "gt": {
      const a = toComparableNumber(actual);
      const b = toComparableNumber(showIf.value);
      return a !== null && b !== null && a > b;
    }
    case "lt": {
      const a = toComparableNumber(actual);
      const b = toComparableNumber(showIf.value);
      return a !== null && b !== null && a < b;
    }
    default:
      return true;
  }
}

/** Resolve a status color from a value + status map, `null` when unmapped. */
export function resolveStatusColor(
  state: unknown,
  statusPath: string | undefined,
  statusMap: BlueprintStatusMap | undefined,
): BlueprintStatusColor | null {
  if (!statusPath || !statusMap) {
    return null;
  }
  const raw = resolvePath(state, statusPath);
  if (raw === undefined || raw === null) {
    return null;
  }
  const key = typeof raw === "string" ? raw : String(raw);
  return statusMap[key] ?? null;
}

const SIZE_TO_TIER: Record<TileSize, BlueprintTier> = {
  "1x1": "compact",
  "1x2": "medium",
  "2x2": "medium",
  "4x2": "wide",
  "2x4": "tall",
  "4x4": "large",
};

// Nearest-smaller fallback chains; `medium` is always defined so every chain ends there.
const TIER_FALLBACK: Record<BlueprintTier, BlueprintTier[]> = {
  compact: ["compact", "medium"],
  medium: ["medium"],
  wide: ["wide", "medium"],
  tall: ["tall", "medium"],
  large: ["large", "wide", "tall", "medium"],
};

export function sizeToTier(size: TileSize): BlueprintTier {
  return SIZE_TO_TIER[size] ?? "medium";
}

export interface ResolvedLayout {
  tier: BlueprintTier;
  node: BlueprintNode;
}

/** Map a widget size to a concrete layout node, applying nearest-smaller fallback. */
export function resolveLayout(blueprint: Blueprint, size: TileSize): ResolvedLayout | null {
  const tier = sizeToTier(size);
  const layouts = blueprint.layouts;
  if (!layouts || typeof layouts !== "object") {
    return null;
  }
  for (const candidate of TIER_FALLBACK[tier]) {
    const node = layouts[candidate];
    if (node) {
      return { tier: candidate, node };
    }
  }
  return layouts.medium ? { tier: "medium", node: layouts.medium } : null;
}

/**
 * Light structural validation of a raw blueprint document. Returns the typed
 * blueprint or `null` when the shape is unusable (so the renderer can show a
 * muted fallback instead of crashing). Full schema validation happens on the
 * backend at install time.
 */
export function parseBlueprint(raw: unknown): Blueprint | null {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    return null;
  }
  const candidate = raw as Record<string, unknown>;
  if (candidate.blueprint_version !== "1") {
    return null;
  }
  const layouts = candidate.layouts;
  if (!layouts || typeof layouts !== "object" || Array.isArray(layouts)) {
    return null;
  }
  const medium = (layouts as Record<string, unknown>).medium;
  if (!medium || typeof medium !== "object") {
    return null;
  }
  return candidate as unknown as Blueprint;
}

File: apps/api/gremlinboard_api/schemas/blueprint.py
from __future__ import annotations

from collections.abc import Iterable
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, model_validator


DotPath = Annotated[
    str,
    Field(pattern=r"^[a-zA-Z_][a-zA-Z0-9_]*(\[[0-9]+\])?(\.[a-zA-Z_][a-zA-Z0-9_]*(\[[0-9]+\])?)*$"),
]
StatusColor = Literal["critical", "warn", "ok", "neutral"]


class BlueprintModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ShowIf(BlueprintModel):
    path: DotPath
    op: Literal["exists", "eq", "gt", "lt"]
    value: Any | None = None


class NodeBase(BlueprintModel):
    show_if: ShowIf | None = None


class LayoutNodeBase(NodeBase):
    gap: Literal["none", "sm", "md"] | None = None
    children: list[BlueprintNode] = Field(min_length=1)


class StackNode(LayoutNodeBase):
    type: Literal["stack"]


class RowNode(LayoutNodeBase):
    type: Literal["row"]


class GridNode(LayoutNodeBase):
    type: Literal["grid"]
    columns: int = Field(ge=2, le=4)


class ScrollNode(LayoutNodeBase):
    type: Literal["scroll"]


class StatusMappedNode(NodeBase):
    status_path: DotPath | None = None
    status_map: dict[str, StatusColor] | None = None


class StatNode(StatusMappedNode):
    type: Literal["stat"]
    label: str = Field(min_length=1)
    value_path: DotPath
    unit: str | None = None
    emphasis: Literal["primary", "secondary"] | None = None
    trend_path: DotPath | None = None


class TextNode(NodeBase):
    type: Literal["text"]
    value_path: DotPath | None = None
    literal: str | None = None
    variant: Literal["title", "body", "caption", "mono"]

    @model_validator(mode="after")
    def validate_text_source(self) -> TextNode:
        if (self.value_path is None) == (self.literal is None):
            raise ValueError("text node must define exactly one of value_path or literal")
        return self


class BadgeItem(BlueprintModel):
    label_path: DotPath | None = None
    literal: str | None = None
    status_path: DotPath | None = None
    status_map: dict[str, StatusColor] | None = None

    @model_validator(mode="after")
    def validate_label_source(self) -> BadgeItem:
        if (self.label_path is None) == (self.literal is None):
            raise ValueError("badge item must define exactly one of label_path or literal")
        return self


class BadgeRowNode(NodeBase):
    type: Literal["badge_row"]
    items: list[BadgeItem] = Field(min_length=1)


class ListItem(BlueprintModel):
    primary_path: DotPath
    secondary_path: DotPath | None = None
    meta_path: DotPath | None = None
    href_path: DotPath | None = None
    status_path: DotPath | None = None
    status_map: dict[str, StatusColor] | None = None


class ListNode(NodeBase):
    type: Literal["list"]
    items_path: DotPath
    limit: int | None = Field(default=None, ge=1)
    item: ListItem


class TableColumn(BlueprintModel):
    header: str = Field(min_length=1)
    value_path: DotPath
    align: Literal["left", "center", "right"] | None = None


class TableNode(NodeBase):
    type: Literal["table"]
    items_path: DotPath
    limit: int | None = Field(default=None, ge=1)
    columns: list[TableColumn] = Field(min_length=1)


class KeyValueEntry(BlueprintModel):
    label: str = Field(min_length=1)
    value_path: DotPath


class KeyValueNode(NodeBase):
    type: Literal["key_value"]
    entries: list[KeyValueEntry] | None = Field(default=None, min_length=1)
    entries_path: DotPath | None = None

    @model_validator(mode="after")
    def validate_entries_source(self) -> KeyValueNode:
        if (self.entries is None) == (self.entries_path is None):
            raise ValueError("key_value node must define exactly one of entries or entries_path")
        return self


class ProgressNode(NodeBase):
    type: Literal["progress"]
    value_path: DotPath
    max_path: DotPath | None = None
    max_literal: float | None = None
    label: str | None = None

    @model_validator(mode="after")
    def validate_max_source(self) -> ProgressNode:
        if self.max_path is not None and self.max_literal is not None:
            raise ValueError("progress node must not define both max_path and max_literal")
        return self


class SparklineNode(NodeBase):
    type: Literal["sparkline"]
    values_path: DotPath
    label: str | None = None


class TimerNode(NodeBase):
    type: Literal["timer"]
    target_path: DotPath
    direction: Literal["down", "up"]
    label_path: DotPath | None = None


class EmptyStateNode(NodeBase):
    type: Literal["empty_state"]
    message: str = Field(min_length=1)
    show_if_empty_path: DotPath


class ActionButtonNode(NodeBase):
    type: Literal["action_button"]
    label: str = Field(min_length=1)
    action: Literal["refresh", "config_patch"]
    config_patch: dict[str, Any] | None = None
    style: Literal["primary", "secondary"] | None = None

    @model_validator(mode="after")
    def validate_config_patch(self) -> ActionButtonNode:
        if self.action == "config_patch" and not self.config_patch:
            raise ValueError("config_patch action requires a non-empty config_patch")
        if self.action == "refresh" and self.config_patch is not None:
            raise ValueError("config_patch must be None for refresh action")
        return self


BlueprintNode = Annotated[
    StackNode
    | RowNode
    | GridNode
    | ScrollNode
    | StatNode
    | TextNode
    | BadgeRowNode
    | ListNode
    | TableNode
    | KeyValueNode
    | ProgressNode
    | SparklineNode
    | TimerNode
    | EmptyStateNode
    | ActionButtonNode,
    Field(discriminator="type"),
]


class BlueprintLayouts(BlueprintModel):
    compact: BlueprintNode | None = None
    medium: BlueprintNode
    wide: BlueprintNode | None = None
    tall: BlueprintNode | None = None
    large: BlueprintNode | None = None


class Blueprint(BlueprintModel):
    blueprint_version: Literal["1"]
    widget_id: str = Field(min_length=1)
    layouts: BlueprintLayouts
    defaults: dict[str, BlueprintNode] | None = None


def validate_blueprint(data: dict[str, Any]) -> Blueprint:
    try:
        return Blueprint.model_validate(data)
    except ValidationError as exc:
        details = "; ".join(f"{'.'.join(str(part) for part in error['loc'])}: {error['msg']}" for error in exc.errors())
        raise ValueError(f"invalid widget blueprint: {details}") from exc


def collect_binding_paths(blueprint: Blueprint) -> set[str]:
    paths: set[str] = set()

    def add(value: str | None) -> None:
        if value is not None:
            paths.add(value)

    def collect_from_show_if(show_if: ShowIf | None) -> None:
        if show_if is not None:
            add(show_if.path)

    def collect_from_nodes(nodes: Iterable[BlueprintNode | None]) -> None:
        for node in nodes:
            if node is not None:
                collect_from_node(node)

    def collect_from_node(node: BlueprintNode) -> None:
        collect_from_show_if(node.show_if)
        if isinstance(node, LayoutNodeBase):
            collect_from_nodes(node.children)
        elif isinstance(node, StatNode):
            add(node.value_path)
            add(node.trend_path)
            add(node.status_path)
        elif isinstance(node, TextNode):
            add(node.value_path)
        elif isinstance(node, BadgeRowNode):
            for item in node.items:
                add(item.label_path)
                add(item.status_path)
        elif isinstance(node, ListNode):
            add(node.items_path)
            add(node.item.primary_path)
            add(node.item.secondary_path)
            add(node.item.meta_path)
            add(node.item.href_path)
            add(node.item.status_path)
        elif isinstance(node, TableNode):
            add(node.items_path)
            for column in node.columns:
                add(column.value_path)
        elif isinstance(node, KeyValueNode):
            add(node.entries_path)
            for entry in node.entries or []:
                add(entry.value_path)
        elif isinstance(node, ProgressNode):
            add(node.value_path)
            add(node.max_path)
        elif isinstance(node, SparklineNode):
            add(node.values_path)
        elif isinstance(node, TimerNode):
            add(node.target_path)
            add(node.label_path)
        elif isinstance(node, EmptyStateNode):
            add(node.show_if_empty_path)

    collect_from_nodes(
        [
            blueprint.layouts.compact,
            blueprint.layouts.medium,
            blueprint.layouts.wide,
            blueprint.layouts.tall,
            blueprint.layouts.large,
        ]
    )
    if blueprint.defaults:
        collect_from_nodes(blueprint.defaults.values())
    return paths


for model in (
    StackNode,
    RowNode,
    GridNode,
    ScrollNode,
    StatNode,
    TextNode,
    BadgeRowNode,
    ListNode,
    TableNode,
    KeyValueNode,
    ProgressNode,
    SparklineNode,
    TimerNode,
    EmptyStateNode,
    ActionButtonNode,
    BlueprintLayouts,
    Blueprint,
):
    model.model_rebuild()

